### DISCUSSION TITLE
Add host type collection

### DISCUSSION THREAD

**Discussion by OuterHeaven07** [ClearFoundry/ClearScript#273] [Q&A]
Hi 

I have a question regarding the embedding of whole type collections.

The examples show how to insert these "standard" collections:

`using Microsoft.ClearScript;`
  `var typeCollection = new HostTypeCollection("mscorlib", "System", "System.Core");`
`engine.AddHostObject("clr", typeCollection);`

Now I include in my application an external library with various dll's (EXT1.dll, EXT2.dll ....) as references. 

Within these external dll's many types are provided, which I want to access in the JS. 

But when I want to add these collections, it throws an exception:
 (- $exception {"The length must not be less than 0 (zero).\r\nParametername: length"}	System.ArgumentOutOfRangeException)

  `var typeCollection = new HostTypeCollection("mscorlib", "System", "System.Core", "EXT1","EXT2");`
`engine.AddHostObject("clr", typeCollection);`

I can't find any other examples on the subject either. Is there any possibility at all to include these types or do they have to be defined individually per 
`engine.AddHostType("....", typeof(<T>));`

Thank you

**Comment by ClearScriptLib** [comment]
Hi @OuterHeaven07,

Can you post the exception stack trace?

Thanks!

**Comment by OuterHeaven07** [comment]
Hi @ClearScriptLib, 

here is the stack trace of the exception: 

   bei System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   bei System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   bei System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean forIntrospection)
   bei System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, Boolean forIntrospection)
   bei System.Reflection.Assembly.Load(String assemblyString)
   bei Microsoft.ClearScript.HostTypeCollection.AddAssembly(String assemblyName)
   bei System.Array.ForEach[T](T[] array, Action`1 action)
   bei {my Application} in .....\Main.cs:Zeile 150.

**Reply by ClearScriptLib** [reply]
Hi @OuterHeaven07,

Thanks for posting the stack trace. Unfortunately it didn't reveal anything obvious. Are you sure the .NET runtime doesn't handle this exception internally?

In any case, we no longer recommend adding assemblies by name. It's one of ClearScript's oldest features, and we've retained it for backward compatibility, but changes in the various .NET runtimes have had negative impact on the feature's reliability, especially for custom assemblies.

As an alternative, consider preloading your custom assemblies by referencing them from your .NET code. Suppose you have two custom assemblies – Ext1.dll and Ext2.dll – containing the types `MyType1` and `MyType2` respectively. Instead of doing this:

``` C#
var typeCollection = new HostTypeCollection("mscorlib", "System", "System.Core", "Ext1", "Ext2");
engine.AddHostObject("clr", typeCollection);
```

Try the following:

``` C#
var typeCollection = new HostTypeCollection("mscorlib", "System", "System.Core");
typeCollection.AddAssembly(typeof(MyType1).Assembly);
typeCollection.AddAssembly(typeof(MyType2).Assembly);
engine.AddHostObject("clr", typeCollection);
```

Please let us know if this helps. Thanks!

**Comment by OuterHeaven07** [comment]
Hi @ClearScriptLib , 

This was exactly what I was looking for 
Thanks a lot
